util/src/matrix.rs

```diff
@@ -0,0 +1,527 @@
+use std::ops::{Index, IndexMut};
+use std::ptr;
+
+/// Calculates the number of elements in a matrix with the given height and width, panicking on
+/// overflow.
+fn size_checked(height: usize, width: usize) -> usize {
+    height.checked_mul(width).expect("memory overflow")
+}
+
+/// Returns an uninitialized boxed slice of lenth `len`.
+fn uninit_box<T>(len: usize) -> Box<[MaybeUninit<T>]> {
+    let mut buf: Vec<MaybeUninit<T>> = Vec::with_capacity(len);
+    debug_assert_eq!(buf.capacity(), len);
+    unsafe {
+        // SAFETY: `with_capacity` guarantees that `buf.capacity() == len`, so we're not violating
+        // memory rules. `MaybeUninit` doesn't need initialization, so the values are already valid.
+        buf.set_len(len);
+    }
+    buf.into_boxed_slice()
+}
+
+/// A non-resizeable matrix.
+pub struct Matrix<T> {
+    data: Box<[T]>,
+    height: usize,
+    width: usize,
+}
+
+impl<T> Matrix<T> {
+    /// Creates a `Matrix<T>` from a boxed slice. The boxed slice is interpreted as a storing the
+    /// matrix in a row-major order. In other words, `data` is split into contiguous chunks, with
+    /// each chunk treated as a row of the matrix.
+    pub fn from_flat_box(height: usize, width: usize, data: Box<[T]>) -> Self {
+        // NB: We ask for the height _and_ the width to handle edge cases when `data.len() == 0`.
+
+        let size = size_checked(height, width);
+        assert_eq!(
+            size,
+            data.len(),
+            "matrix dimensions do not match boxed slice length"
+        );
+
+        Self {
+            data,
+            height,
+            width,
+        }
+    }
+
+    /// Creates a `Matrix<T>` from a vector. The vector is interpreted as a storing the matrix in a
+    /// row-major order. In other words, `v` is split into contiguous chunks, with each chunk
+    /// treated as a row of the matrix.
+    ///
+    /// This is a convenience wrapper of `from_flat_box`.
+    pub fn from_flat_vec(height: usize, width: usize, v: Vec<T>) -> Self {
+        Self::from_flat_box(height, width, v.into_boxed_slice())
+    }
+
+    /// Creates an uninitialized matrix of specified height and width.
+    pub fn new_uninit(height: usize, width: usize) -> Matrix<MaybeUninit<T>> {
+        let size = size_checked(height, width);
+        let buf_box = uninit_box::<T>(size);
+        Matrix::from_flat_box(height, width, buf_box)
+    }
+
+    /// Creates a new matrix with a specified height and width, initialized with a provided closure.
+    ///
+    /// This function allocates an uninitialized matrix of with a given `height` and `width`. A
+    /// reference to this uninitialized array is passed to the user-provided closure, which is
+    /// responsible for initializing all elements.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as `f` must initialization on every element of the matrix.
+    pub unsafe fn new_with<F: FnOnce(&mut Matrix<MaybeUninit<T>>)>(
```

Comment:
For a public API I'd expect (in the principle-of-least-surprise sense):

```rust
/// Creates a new matrix with a specified height and width, initialized with a provided closure.
///
/// The closure will be called in row-major order for each element.
pub fn new_with<F: FnMut(usize, usize) -> T>(height: usize, width: usize, mut f: F) -> Self {
    let mut res = Self::new_uninit(height, width);
    for (i, row) in res.iter_mut().enumerate() {
        for (j, element) in row.iter_mut().enumerate() {
            element.write(f(i, j));
        }
    }
    unsafe {
        // SAFETY: All elements are now initialized.
        res.assume_init()
    }
}
```

It's obviously not as flexible as the original, but it covers quite a few use-cases and is simple and safe:

```rust
#[test]
fn test_new_with() {
    let m = Matrix::new_with(HEIGHT, WIDTH, |i, j| (WIDTH * i + j) as u64);
    check_matrix(m);
}
```

If you do need the old behaviour the existing public APIs already allow that cleanly

```rust
#[test]
fn test_new_with3() {
    let m = {
        let mut m = Matrix::new_uninit(HEIGHT, WIDTH);
        for i in 0..HEIGHT {
            for j in 0..WIDTH {
                m[i][j].write((WIDTH * i + j) as u64);
            }
        }
        unsafe { m.assume_init() }
    };
    check_matrix(m);
}
```

Reply:
You're right that it would be safer, but it also forces a particular evaluation order that's not always optimal (e.g. if you're doing a transpose). I don't think this constructor is actually needed for anything, so I'll just get rid of it and we can bring it back later.